pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1087 @@
+	///   fully, which we denote by adding the 500 to the `foreign_done` amount.
+	///   Moreover, we can immediately invest the 500. The resulting state is
+	///   `(done_amount = 1500, investing = 500)`.
+	/// * b) If we increase by 400, we can reduce the `foreign_swap.amount` only
+	///   by 400 and increase both the `investing` as well as `foreign_done`
+	///   amount by that. The resulting state is
+	/// `(done_amount = 1400, foreign_swap.amount = 100, investing = 400)`.
+	/// * c) If we increase by 600, we can reduce the `foreign_swap.amount`
+	///   fully and need to add a swap into pool currency for 100. Moreover both
+	///   the `investing` as well as `foreign_done` amount can only be increased
+	///   by 500. The resulting state is
+	/// `(done_amount = 1500, pool_swap.amount = 100, investing = 500)`.
+	///
+	/// NOTES:
+	/// * We can never directly compare `swap.amount` and `invest_amount` with
+	///   `foreign_swap.amount` and `done_amount` if the currencies mismatch as
+	///   the former pair is denominated in pool currency and the latter one in
+	///   foreign currency.
+	/// * We can ignore handling all states which include `*SwapIntoForeignDone`
+	///   without `ActiveSwapIntoForeignCurrency*` as we consume the done amount
+	///   and transition in the post transition phase. To be safe and to not
+	///   make any unhandled assumptions, we throw `DispatchError::Other` for
+	///   these states though we need to make sure this can never occur!
+	fn handle_increase(
+		&self,
+		swap: Swap<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		if swap.currency_in == swap.currency_out {
+			return Self::handle_increase_non_foreign(self, swap);
+		}
+
+		match &self {
+			Self::NoState => Ok(Self::ActiveSwapIntoPoolCurrency { swap }),
+			// Add pool swap
+			Self::InvestmentOngoing { invest_amount } => {
+				Ok(Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+					swap,
+					invest_amount: *invest_amount,
+				})
+			}
+			// Bump pool swap
+			Self::ActiveSwapIntoPoolCurrency { swap: pool_swap } => {
+				swap.ensure_currencies_match(pool_swap, true)?;
+				Ok(Self::ActiveSwapIntoPoolCurrency {
+					swap: Swap {
+						amount: swap.amount.ensure_add(pool_swap.amount)?,
+						..swap
+					},
+				})
+			}
+			// Reduce foreign swap amount by the increasing amount and increase investing amount as
+			// well adding foreign_done amount by the minimum of active swap amounts
+			Self::ActiveSwapIntoForeignCurrency { swap: foreign_swap } => {
+				swap.ensure_currencies_match(foreign_swap, false)?;
+				let foreign_amount_pool_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_in,
+					swap.currency_out,
+					foreign_swap.amount,
+				)?;
+				let pool_amount_foreign_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_out,
+					swap.currency_in,
+					swap.amount,
+				)?;
+
+				match swap.amount.cmp(&foreign_amount_pool_denominated) {
```

Comment:
Supposing the `swap.amount` is a foreign swap, that amount is measured as foreign currency, which can not be compared to `foreign_amount_pool_denominated` (measured as pool currency).

Am I wrong?

Reply:
Autoanswer: Do not apply because it's checked just before by `ensure_currencies_match` 😄